fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIndexPlan.java

```diff
@@ -69,48 +76,98 @@
                                  @Nonnull final Optional<ValueIndexScanMatchCandidate> matchCandidateOptional) {
         this.indexName = indexName;
+        this.commonPrimaryKey = commonPrimaryKey;
         this.scanParameters = scanParameters;
+        this.useIndexPrefetch = useIndexPrefetch;
         this.reverse = reverse;
         this.strictlySorted = strictlySorted;
         this.matchCandidateOptional = matchCandidateOptional;
     }
 
+    @Nonnull
+    @Override
+    public <M extends Message> RecordCursor<QueryResult> executePlan(@Nonnull final FDBRecordStoreBase<M> store, @Nonnull final EvaluationContext context,
+                                                                     @Nullable final byte[] continuation, @Nonnull final ExecuteProperties executeProperties) {
+        switch (useIndexPrefetch) {
+            case NONE:
+                // Use the default implementation for index scan
+                return RecordQueryPlanWithIndex.super.executePlan(store, context, continuation, executeProperties);
+            case USE_INDEX_PREFETCH:
+                // Use index prefetch without fallback
+                return executeUsingIndexPrefetch(store, context, continuation, executeProperties);
+            case USE_INDEX_PREFETCH_WITH_FALLBACK:
+                // Use Index prefetch and fall back to regular index scan.
+                // This is not ideal, but using teh fallback mechanism here separates the execution part from the planning part
+                // (No need to plan again) and from failures at other parts of the execution.
+                try {
+                    return executeUsingIndexPrefetch(store, context, continuation, executeProperties);
+                } catch (Exception ex) {
+                    KeyValueLogMessage message = KeyValueLogMessage.build("Index Prefetch plan failed, falling back to Index scan",
+                            LogMessageKeys.PLAN_HASH, planHash(PlanHashKind.STRUCTURAL_WITHOUT_LITERALS));
+                    LOGGER.error(message.toString(), ex);
+                    return RecordQueryPlanWithIndex.super.executePlan(store, context, continuation, executeProperties);
+                }
+            default:
+                throw new RecordCoreException("Unknown useIndexPrefetch option").addLogInfo("option", useIndexPrefetch);
+        }
+    }
+
+    @Nonnull
+    private <M extends Message> RecordCursor<QueryResult> executeUsingIndexPrefetch(@Nonnull final FDBRecordStoreBase<M> store, @Nonnull final EvaluationContext context,
+                                                                                    @Nullable final byte [] continuation, @Nonnull final ExecuteProperties executeProperties) {
+        final TupleRange range = getComparisons().toTupleRange(store, context);
+        return store.scanIndexPrefetch(getIndexName(), range, getCommonPrimaryKey(), continuation, executeProperties.asScanProperties(isReverse()))
```

Comment:
`scanIndexPrefetch` is an alternative for `scanIndexRecords`, except that because of `executeEntries` alternatives, the plan splits that up into `scanIndex` and `fetchIndexRecords`. I think some refactoring of that could avoid more duplication.

Reply:
Just to make sure I understand what you have in mind: Do you mean that the `RecordQueryPlanWithIndex.executePlan()` would always be the one executing the scan, and that the overridden `executeEntries()` make the decision on which FDB method to call? And that then the `store.fetchIndexRecords()` either use the existing fetch method or create the record from the `MappedKeyValue` based on the type of the entry in the cursor?

Reply:
Discussed and I believe resolved.

Reply:
Yes.